fe/fe-core/src/main/java/com/starrocks/sql/optimizer/transformer/WindowTransformer.java

```diff
@@ -296,18 +300,71 @@ public static List<WindowTransformer.PartitionGroup> reorderWindowOperator(
                     .setOrderByElements(orderings)
                     .setAnalyticWindow(windowOperator.getWindow())
                     .setEnforceSortColumns(sortEnforceProperty.stream().distinct().collect(Collectors.toList()))
+                    .setUseHashBasedPartition(windowOperator.useHashBasedPartition)
                     .build());
         }
 
+        List<LogicalWindowOperator> hashBasedWindowOperators = logicalWindowOperators.stream()
+                .filter(LogicalWindowOperator::isUseHashBasedPartition)
```

Comment:
if hash partition can also be a physical property, it would be better, in EnforceAndCost task,  choose hash or sort according to the enforcer's cost estimation.

Reply:
Yes, it should be better that way. But the truth is, the cardinality of multiple-columns is hard to predict gracefully under the current statistics framework. So I choose to use hint by now, it is far beyond the best way but at least offer users a chance to tune performance.